File: backend/utils/public_url.py
from urllib.parse import urlsplit

from config.service_conf import BIZ_PUBLIC_BASE_URL


STORAGE_PREFIXES = (
    "/storage/",
    "/storage_pointcloud/",
    "/storage_mesh/",
    "/storage_gs/",
)


def build_public_url(path: str) -> str:
    normalized_path = "/" + path.lstrip("/")
    if BIZ_PUBLIC_BASE_URL:
        return f"{BIZ_PUBLIC_BASE_URL}{normalized_path}"
    return normalized_path


def normalize_public_url(url: str | None) -> str | None:
    if not url:
        return url

    parsed = urlsplit(url)
    path = parsed.path if parsed.scheme and parsed.netloc else url

    for prefix in STORAGE_PREFIXES:
        if path.startswith(prefix):
            return build_public_url(path)

    return url

File: frontend/src/utils/publicUrl.js
const STORAGE_PREFIXES = [
  '/storage/',
  '/storage_pointcloud/',
  '/storage_mesh/',
  '/storage_gs/',
]

export function resolvePublicUrl(url) {
  if (!url) return url

  try {
    const parsed = new URL(url, window.location.origin)
    const isLocalBackend =
      ['localhost', '127.0.0.1', '0.0.0.0'].includes(parsed.hostname) &&
      ['8000', ''].includes(parsed.port)

    if (isLocalBackend && STORAGE_PREFIXES.some((prefix) => parsed.pathname.startsWith(prefix))) {
      return `${parsed.pathname}${parsed.search}${parsed.hash}`
    }
  } catch {
    // Keep the original value when it is not a URL-like string.
  }

  return url
}
